src/util/desktophelper.cpp

```diff
@@ -0,0 +1,169 @@
+
+#include "desktophelper.h"
+
+#include <QtDebug>
+#include <QString>
+#include <QDesktopServices>
+#include <QDir>
+#include <QUrl>
+#include <QProcess>
+
+#ifdef Q_OS_LINUX
+#include <QtDBus/QtDBus>
+#include <QFileInfo>
+#endif
+
+namespace {
+    const QString kSelectInFreedesktop = "fd";
+    const QString kSelectInXfce = "xf";
+    QString sSelectInFileBrowserCommand;
+
+    QString getSelectInFileBrowserCommand() {
+#if defined(Q_OS_MAC)
+        return "open -R \"%1\"";
+#elif defined(Q_OS_WIN)
+        return "explorer.exe /select,\"%1\"";
+#elif defined(Q_OS_LINUX)
+        QProcess proc;
+        QString output;
+        proc.start("xdg-mime",
+                QStringList() << "query" << "default" << "inode/directory");
+        proc.waitForFinished();
+        output = proc.readLine().simplified();
+        if (output == "kfmclient_dir.desktop") {
+            return "konqueror --select \"%1\"";
+        } else if (output == "Thunar.desktop" ||
+                output == "Thunar-folder-handler.desktop") {
+            return kSelectInXfce; // Thunar has no --select option
+        } else {
+            // Known file manager with org.freedesktop.FileManager1 interface
+            // * Nautilus
+            // * Nemo
+            // * Deepin File Manager
+            // * Swordfish
+            // * Caja
+            // No solution for
+            // * pcmanfm
+            return kSelectInFreedesktop; // no special command, try Freedesktop and fallback to QDesktopServices";
+        }
+#else
+        return ""; // no special command use QDesktopServices";
+#endif
+    }
+
+    QString removeChildDir(const QString& path) {
+        int index = path.lastIndexOf(QChar('/'));
+        return path.left(index);
+    }
+
+#ifdef Q_OS_LINUX
+    bool selectInFreedesktop(const QString& path) {
+        const QUrl fileurl = QUrl::fromLocalFile(path);
+        const QStringList args(fileurl.toString());
+        QDBusMessage msg = QDBusMessage::createMethodCall(
+                "org.freedesktop.FileManager1",
+                "/org/freedesktop/FileManager1",
+                "org.freedesktop.FileManager1",
+                "ShowItems");
+        msg << args << "";
+        const QDBusMessage response = QDBusConnection::sessionBus().call(msg);
+        const bool success = (response.type() != QDBusMessage::MessageType::ErrorMessage);
+        return success;
+    }
+
+    bool selectInXfce(const QString& path) {
+        const QString folder = QFileInfo(path).dir().absolutePath();
```

Comment:
It'd be nice to combine the two temporary QFileInfo's into one here -- IIRC QFileInfo invokes a stat of the filesystem.